base/gmp.jl

```diff
@@ -283,31 +283,21 @@ BigInt(x::Float16) = BigInt(Float64(x))
 BigInt(x::Float32) = BigInt(Float64(x))
 
 function BigInt(x::Integer)
-    if x < 0
-        if typemin(Clong) <= x
-            return BigInt(convert(Clong,x))
-        end
-        b = BigInt(0)
-        shift = 0
-        while x < -1
-            b += BigInt(~UInt32(x&0xffffffff))<<shift
-            x >>= 32
-            shift += 32
-        end
-        return -b-1
-    else
-        if x <= typemax(Culong)
-            return BigInt(convert(Culong,x))
-        end
-        b = BigInt(0)
-        shift = 0
-        while x > 0
-            b += BigInt(UInt32(x&0xffffffff))<<shift
-            x >>>= 32
-            shift += 32
-        end
-        return b
+    x == 0 && return BigInt(Culong(0))
+    nd = ndigits(x, base=2)
+    z = MPZ.realloc2(nd)
+    s = sign(x)
+    s == -1 && (x = -x)
```

Comment:
This fails for `x == typemin(x)`. Fix incoming.

Reply:
really? I get the correct result (note that it uses unsigned shift `>>>` below).

Reply:
to be fair, it might be cleaner to explicitly call `unsigned` here

Reply:
and add `typemin`/`typemax` as test cases.

Reply:
https://github.com/JuliaLang/julia/pull/29340

Reply:
A small consolation for this embarassing bug is that it was [catched](https://github.com/JuliaLang/julia/pull/29238#issuecomment-423942246) by a recent test of mine :sweat_smile: